### ISSUE TITLE
Select, OrderBy. an incorrect sql is generated

### ISSUE THREAD

**Issue by sergeylopatyuk** [linq2db/linq2db#1218]
Hello, I need a query
```sql
SELECT DISTINCT SomeField FROM TableName ORDER BY DateTimeField DESC LIMIT 10
```

I do the following:
```cs
var result = CurrentContext.GetTable<TableName>()
   .Where(specification)
   .OrderByDescending(x => x.DateTimeField)
   .Select(x => x.SomeField)
   .Distinct()
   .Take(10)
   .ToList();
```
or it (because it is impossible to sort on a field which is not in select)
```cs
var result = CurrentContext.GetTable<StatByPlayer>()
   .Where(specification)
   .Select(x => new  { x.SomeField, x.DateTimeField})
   .OrderByDescending(x => x.DateTimeField)
   .Select(x => x.SomeField)
   .Distinct()
   .Take(10)
   .ToList();
```
but I get the following sql:

```sql
SELECT
        t2.SomeField as SomeField
FROM
        (
                SELECT DISTINCT
                        t1.SomeField
                FROM
                        TableName t1
                WHERE
                        t1.UserId = @userId
        ) t2
LIMIT 10
```
**what's my mistake?**

**Comment by sdanyliv** [comment]
@sergeylopatyuk, have you tried your query? 
SQL Server will throw error:
`ORDER BY items must appear in the select list if SELECT DISTINCT is specified.`

If you need this order you can do that by Window Function:
```cs
var filtered = CurrentContext.GetTable<TableName>().Where(specification);

var queryWindowed = from t in filtered
   select new 
   { 
       t.SomeField,
       RN = Sql.Ext.RowNumber().Over().PartitionBy(t.SomeField).OrderByDesc(t.DateTimeField)
   };

var result = queryWindowed
   .Where(x => x.RN == 1)
   .Select(x => x.SomeField)
   .Take(10)
   .ToList();
```

**Comment by sergeylopatyuk** [comment]
Thanks, but sorry, did not specify. I'm using MySql 5.7. There are no Window Function in it.
I had to use the grouping for this query:

```cs
CurrentContext.GetTable<TableName>
    .Where(specification)
    .GroupBy(x => x.SomeField)
    .Select(x => new
    {
           SomeField = x.Key,
           MaxDateForGroup = x.Max(z => z.DateTimeField)
    })
    .OrderByDescending(x => x.MaxDateForGroup)
    .Select(x => x.SomeField)
    .Take(take)
    .ToList();
```
And I get the correct and working Sql:
```sql
SELECT
        `t1`.`SomeField`
FROM
        `TableName` `t1`
WHERE
        `t1`.`UserId` = @userId
GROUP BY
        `t1`.`SomeField`
ORDER BY
        Max(`t1`.`DateTimeField`) DESC
LIMIT 10
```
But this one I think is simpler for the database:
```sql
SELECT DISTINCT SomeField FROM TableName ORDER BY DateTimeField DESC LIMIT 10
```

**Comment by sdanyliv** [comment]
MySql works with this query?
Also, have you compared execution plans?

**Comment by sergeylopatyuk** [comment]
Yes, this sql is working. 
The execution plan seems to be identical. probably this is the answer.

**Comment by sdanyliv** [comment]
@sergeylopatyuk, for sure by adding additional efforts we can support such queries. But i’m not sure that this is not wasting time.
BTW, first time i saw that MySql can optimize something by itself ;)

**Comment by sdanyliv** [comment]
@sergeylopatyuk, i’ll check if it is not too hard to implement, maybe we can support this. Especially because we can transform query to grouping for databases that do not support such syntax.

**Comment by MaceWindu** [comment]
Fixed by #1221
